demos/meltdown_br.cc

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Meltdown-BR demonstrates speculation over the "bound" instruction on x86.
+ * That instruction is allowed only in 32-bit mode, so the example cannot work
+ * on 64-bits. We check with the "bound" instruction whether a given offset
+ * belongs to the public_data string. In the case of safe offset it passes,
+ * when the offset points to the private_data, the bounds check fails and
+ * SIGSEGV is triggered. However, the speculative execution continues past the
+ * bounds check.
+ **/
+
+#if !defined(__linux__) && !defined(__APPLE__)
+#  error Unsupported OS. Linux or MacOS required.
+#endif
+
+#ifndef __i386__
+#  error Unsupported architecture. 32-bit x86 required.
+#endif
+
+#include <array>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+
+#include <signal.h>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+#include "utils.h"
+
+const char *public_data = "Hello, world!";
+const char *private_data = "It's a s3kr3t!!!";
+
+// ICC requires the offset variable to be volatile. If it isn't, ICC schedules
+// to spill it to stack after the second ForceRead call and that never happens
+// because of the SIGSEGV and architectural jumping over that section.
+// In the next loop the restore from stack spill just loads some random value
+// from the stack that was not rewritten.
```

Comment:
It seems like we're just very lucky other compilers haven't broken us in this way yet. It would be good to find a method for error recovery that plays better with the C++ execution model.

Reply:
That's true. I'll check it eventually: https://github.com/google/safeside/issues/74